refactor: remove grunt usage and replace with npm scripts

ewanharris opened the pull request:
This removes the usage of grunt and goes all in on using npm scripts in its place.

Here's a before and after of what you want to run

`grunt`/`grunt lint` -> `npm run lint`
`grunt format:TYPE` -> `npm run format:TYPE` where TYPE is one of android, ios, or js

There is now also separate linting for the each type that can be used like `npm run:TYPE` where type is one of android, docs, ios, or js

I'm filing this as a WIP for now as I'm going to start work on hooking up the existing unit tests under android/cli in this PR also. I just wanted to see how Jenkins handles these changes, and gather some opinions

Edit:

Updated PR to run any tests that match the pattern `**/cli/tests/test-*.js`, this includes the existing AndroidManifest parser tests, which have also been updated to use chai.expect (the existing should usage failed on every `should match object` test as it seems between those being written, and today should also started asserting the prototype)

Next steps is to try integrate those into CI

ewanharris committed:
refactor: remove grunt usage and replace with npm scripts

This commit removes the usage of grunt and goes all in on using npm scripts in its place.

Here's a before and after of what you want to run

grunt/grunt lint -> npm run lint
grunt format:TYPE -> npm run format:TYPE where TYPE is one of android, ios, or js

There is now also separate linting for the each type that can be used like npm run:TYPE
where type is one of android, docs, ios, or js

build commented:
<!--
  1 failure:  Tests have failed...
  1 warning:  There is no linke...
  4 messages
  4 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>There is no linked JIRA ticket in the PR body. Please include the URL of the relevant JIRA ticket. If you need to, you may file a ticket on <a href="https://jira.appcelerator.org/secure/CreateIssue!default.jspa">JIRA</a></td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:thumbsup: Hey!, You deleted more code than you added. That's awesome!</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-11230/1/artifact/dist/mobilesdk-8.3.0.v20190918133522-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 1 tests have failed
There are 1 tests failing and 473 skipped out of 4370 total tests.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.emulator.Titanium.Geolocation</td><td>#forwardGeocoder()</td><td>2.043</td><td><pre>Error: expected 37.3909049 to be approximately 37.387 ±0.002
at Assertion.fail (/node_modules/should/cjs/should.js:275:17)
      at Assertion.value [as approximately] (/node_modules/should/cjs/should.js:356:19)
      at Geolocation.&lt;anonymous&gt; (/ti.geolocation.test.js:259:34)</pre></td></tr>
</table>


Dependencies removed: <a href='https://www.npmjs.com/package/grunt'><code>grunt</code></a>, <a href='https://www.npmjs.com/package/grunt-clang-format'><code>grunt-clang-format</code></a>, <a href='https://www.npmjs.com/package/grunt-cli'><code>grunt-cli</code></a>, <a href='https://www.npmjs.com/package/grunt-contrib-clean'><code>grunt-contrib-clean</code></a> and <a href='https://www.npmjs.com/package/grunt-eslint'><code>grunt-eslint</code></a>.

New dependencies added: <a href='https://www.npmjs.com/package/npm-run-all'><code>npm-run-all</code></a>.


<h2><a href="https://github.com/mysticatea/npm-run-all">npm-run-all</a></h2>
<p>Author: Toru Nagashima</p>
<p>Description: A CLI tool to run multiple npm-scripts in parallel or sequential.</p>
<p>Homepage: <a href="https://github.com/mysticatea/npm-run-all">https://github.com/mysticatea/npm-run-all</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 4 years ago</td></tr><tr><td>Last Updated</td><td>5 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>58</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/ansi-styles'><code>ansi-styles</code></a>, <a href='https://www.npmjs.com/package/chalk'><code>chalk</code></a>, <a href='https://www.npmjs.com/package/cross-spawn'><code>cross-spawn</code></a>, <a href='https://www.npmjs.com/package/memorystream'><code>memorystream</code></a>, <a href='https://www.npmjs.com/package/minimatch'><code>minimatch</code></a>, <a href='https://www.npmjs.com/package/pidtree'><code>pidtree</code></a>, <a href='https://www.npmjs.com/package/read-pkg'><code>read-pkg</code></a>, <a href='https://www.npmjs.com/package/shell-quote'><code>shell-quote</code></a> and <a href='https://www.npmjs.com/package/string.prototype.padend'><code>string.prototype.padend</code></a></td></tr><tr><td>Keywords</td><td>cli, command, commandline, tool, npm, npm-scripts, run, sequential, serial, parallel and task</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

| index | [npm-run-all] | [run-s] | [run-p] | [Node API] |
|-------|---------------|---------|---------|------------|

# npm-run-all

[![npm version](https://img.shields.io/npm/v/npm-run-all.svg)](https://www.npmjs.com/package/npm-run-all)
[![Downloads/month](https://img.shields.io/npm/dm/npm-run-all.svg)](http://www.npmtrends.com/npm-run-all)
[![Build Status](https://travis-ci.org/mysticatea/npm-run-all.svg?branch=master)](https://travis-ci.org/mysticatea/npm-run-all)
[![Build status](https://ci.appveyor.com/api/projects/status/v0owd44q1r7hceir/branch/master?svg=true)](https://ci.appveyor.com/project/mysticatea/npm-run-all/branch/master)
[![Coverage Status](https://codecov.io/gh/mysticatea/eslint-plugin-node/branch/master/graph/badge.svg)](https://codecov.io/gh/mysticatea/npm-run-all)
[![Dependency Status](https://david-dm.org/mysticatea/npm-run-all.svg)](https://david-dm.org/mysticatea/npm-run-all)

A CLI tool to run multiple npm-scripts in parallel or sequential.

## ⤴️ Motivation

- **Simplify.** The official `npm run-script` command cannot run multiple scripts, so if we want to run multiple scripts, it's redundant a bit. Let's shorten it by glob-like patterns.<br>
  Before: `npm run clean && npm run build:css && npm run build:js && npm run build:html`<br>
  After: `npm-run-all clean build:*`
- **Cross platform.** We sometimes use `&` to run multiple command in parallel, but `cmd.exe` (`npm run-script` uses it by default) does not support the `&`. Half of Node.js users are using it on Windows, so the use of `&` might block contributions. `npm-run-all --parallel` works well on Windows as well.

## 💿 Installation

```bash
$ npm install npm-run-all --save-dev
# or
$ yarn add npm-run-all --dev
```

- It requires `Node@>=4`.

## 📖 Usage

### CLI Commands

This `npm-run-all` package provides 3 CLI commands.

- [npm-run-all]
- [run-s]
- [run-p]

The main command is [npm-run-all].
We can make complex plans with [npm-run-all] command.

Both [run-s] and [run-p] are shorthand commands.
[run-s] is for sequential, [run-p] is for parallel.
We can make simple plans with those commands.

#### Yarn Compatibility

If a script is invoked with Yarn, `npm-run-all` will correctly use Yarn to execute the plan's child scripts.

### Node API

This `npm-run-all` package provides Node API.

- [Node API]

## 📰 Changelog

- https://github.com/mysticatea/npm-run-all/releases

## 🍻 Contributing

Welcome♡

### Bug Reports or Feature Requests

Please use GitHub Issues.

### Correct Documents

Please use GitHub Pull Requests.

I'm not familiar with English, so I especially thank you for documents' corrections.

### Implementing

Please use GitHub Pull Requests.

There are some npm-scripts to help developments.

- **npm test** - Run tests and collect coverage.
- **npm run clean** - Delete temporary files.
- **npm run lint** - Run ESLint.
- **npm run watch** - Run tests (not collect coverage) on every file change.

[npm-run-all]: docs/npm-run-all.md
[run-s]: docs/run-s.md
[run-p]: docs/run-p.md
[Node API]: docs/node-api.md


</details>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4d3b1c529fb9be54f4b6113ee8eaac85228cf86e
</p>

ewanharris committed:
test: run tests under cli folders

Configure mocha to run any test-*.js files that exist under <platform>/cli/tests with coverage. Also refactor te existing androidmanifest tests to use chai over should

build commented on android/cli/tests/test-androidmanifest.js:
```diff
@@ -40,10 +41,11 @@ describe('AndroidManifest', function () {
 
 	describe('<application>', function () {
-		var am = new AndroidManifest(path.resolve('./resources/AndroidManifest_application.xml'));
+		const manifest = path.join(__dirname, './resources/AndroidManifest_application.xml');
```
<!--
  0 failure: 
  1 warning:  android/cli/tests...
  
  
  DangerID: danger-id-Danger;
  File: android/cli/tests/test-androidmanifest.js;
  Line: 44;
-->

- :warning: android/cli/tests/test-androidmanifest.js line 44 – 'manifest' is assigned a value but never used. (no-unused-vars)

sgtcoolguy commented on package.json:
```diff
@@ -18,13 +18,23 @@
+    "format:android": "./build/scons format --android --fix",
+    "format:ios": "./build/scons format --ios --fix",
+    "format:js": "npm run lint:js -- --fix",
+    "unit-tests": "nyc --exclude **/cli/tests/test-*.js mocha **/cli/tests/test-*.js"
```
so in keeping with this theme, I would suggest we name this "test:unit" and migrate "android-test" and "ios-test" to be "test:android" and "test:ios" respectively.

sgtcoolguy replied:
or maybe even "test:cli"? "test:js" hopefully align the naming across the various actions.

ewanharris replied:
Added the following `test:android`, `test:cli`, `test:ios`.

I didn't point `npm test` to run these though as I think that is probably unwanted. I also wasn't sure on whether to have `cli` or `js` in the naming, I went with the former as while it isn't inline with the format/lint steps, it is clearer as to what the task does.

ewanharris committed:
refactor: move test scripts to common format

ewanharris committed:
ci: run cli unit tests in jenkins

ewanharris committed:
ci: stash and unstash required files for cli tests

ewanharris commented on Jenkinsfile:
```diff
@@ -168,6 +168,27 @@ def unitTests(os, nodeVersion, npmVersion, testSuiteBranch, testOnDevices) {
 
+def cliUnitTests(nodeVersion, npmVersion) {
+	return {
+		node('git && osx') { // ToDo: refactor to try and run across mac, linux, and windows?
```
I guess this may come at the annoyance of having to skip certain tests on certain platforms

janvennemann reviewed (approved):
Awesome, @ewanharris! Looks good to me!

sgtcoolguy commented:
manually rebased on master, handled conflicts and merged.
